hera_pspec/container.py

```diff
@@ -23,24 +52,63 @@ def __init__(self, filename, mode='r'):
             Whether to load the HDF5 file as read/write ('rw') or read-only
             ('r'). If 'rw' is specified and the file doesn't exist, an empty
             one will be created.
+        
+        keep_open : bool, optional
+            Whether the HDF5 file should be kept open, or opened and then 
+            closed again each time an operation is performed. Setting 
+            keep_open=False is helpful for multi-process access patterns. 
+            Default: True (keep file open).
         """
         self.filename = filename
+        self.keep_open = keep_open
         self.mode = mode
         if mode not in ['r', 'rw']:
             raise ValueError("Must set mode to either 'r' or 'rw'.")
 
-        # Open file ready for reading and/or writing
+        # Open file ready for reading and/or writing (if not in transactional mode)
         self.data = None
-        self._open()
-
+        if keep_open: self._open()
+    
+    
     def _open(self):
         """
-        Open HDF5 file ready for reading/writing.
+        Open HDF5 file ready for reading/writing. Does nothing if the file is 
+        already open.
+        
+        This method uses HDF5's single writer, multiple reader (swmr) mode, 
+        which allows multiple handles to exist for the same file at the same 
+        time, as long as only one is in rw mode. The rw instance should be the 
+        *first* one that is created; if a read-only instance is already open 
+        when a rw instance is created, an error will be raised by h5py. 
         """
+        if self.data is not None: return
+        
         # Convert user-specified mode to a mode that HDF5 recognizes. We only
         # allow non-destructive operations!
         mode = 'a' if self.mode == 'rw' else 'r'
-        self.data = h5py.File(self.filename, mode)
+        swmr = True if self.mode == 'r' else False
```

Comment:
Here's a case where I think it'd be much more obvious to break this out into a traditional multi-line if/else block. The line above also uses the ternary conditional, which (for me) is hard to parse. You might even be able to handle these different cases within a single if/elif/else series of conditionals.